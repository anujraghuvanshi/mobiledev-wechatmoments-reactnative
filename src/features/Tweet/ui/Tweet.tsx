import React, {ReactElement, useState} from 'react';
import {
  Image,
  StyleSheet,
  View,
  Text,
  ViewStyle,
  TextStyle,
} from 'react-native';

import {BasicStyle, ITweet} from '../../../types';
import {TouchableOpacity} from 'react-native-gesture-handler';

interface ITweetProps {
  tweet: ITweet;
}

export function Tweet({tweet}: ITweetProps): ReactElement {
  const [showAllImages, setshowAllImages] = useState<boolean>(
    tweet.images && tweet?.images.length <= 4 ? true : false,
  );
  const [isReadingMore, setisReadingMore] = useState(false);

  return (
    <View testID="tweet-wrapper" style={styles.container}>
      {tweet.error || tweet['unknown error'] ? (
        <Text testID="error-text">
          Tweet with error: {tweet.error || tweet['unknown error']}{' '}
        </Text>
      ) : (
        <>
          <Image
            style={styles.image}
            testID="sender-image"
            source={{
              uri: tweet?.sender?.avatar || '',
              width: 40,
              height: 40,
            }}
          />
          <View style={styles.tweetContainer}>
            <View>
              <Text style={styles.sender}>
                {tweet?.sender?.nick || tweet?.sender?.username}
              </Text>
              {tweet?.content && (
                <View style={styles.contentWrapper}>
                  <Text
                    testID="tweet-content"
                    style={styles.text}
                    numberOfLines={isReadingMore ? undefined : 2}>
                    {tweet.content}
                  </Text>
                  <Text
                    style={styles.readMore}
                    testID="read-more-trigger"
                    onPress={() => {
                      setisReadingMore(!isReadingMore);
                    }}>
                    {isReadingMore ? 'Read Less' : 'Read more'}
                  </Text>
                </View>
              )}
            </View>
            {tweet?.images?.length && (
              <View style={styles.tweetImageContainer}>
                <View
                  testID="tweet-images-wrapper"
                  style={styles.imagesWrapper}>
                  {(showAllImages
                    ? tweet.images
                    : tweet.images.slice(0, 4)
                  ).map((image, index) => (
                    <Image
                      testID="tweet-image"
                      key={`image-${image?.url?.split('/').pop()}-${index}`}
                      style={styles.image}
                      source={{
                        uri: image?.url || '',
                        width: 64,
                        height: 64,
                      }}
                    />
                  ))}
                </View>
                {tweet?.images?.length > 4 && (
                  <TouchableOpacity
                    onPress={() => {
                      setshowAllImages(!showAllImages);
                    }}>
                    <Text>Show All Images</Text>
                  </TouchableOpacity>
                )}
              </View>
            )}
            {tweet.comments && (
              <View testID="tweet-comments-wrapper" style={styles.commentsWrap}>
                {tweet.comments.map((comment, index) => (
                  <View
                    key={`${index}-${comment.sender.username}`}
                    style={styles.comment}>
                    <Text style={styles.commentSender}>
                      {comment.sender.nick}:{' '}
                    </Text>
                    <Text>{comment.content}</Text>
                  </View>
                ))}
              </View>
            )}
          </View>
        </>
      )}
    </View>
  );
}

interface AdditionalStyle {
  imagesWrapper: ViewStyle;
  tweetContainer: ViewStyle;
  sender: TextStyle;
  commentsWrap: ViewStyle;
  comment: ViewStyle;
  commentSender: TextStyle;
  tweetImageContainer: ViewStyle;
  contentWrapper: ViewStyle;
  readMore: TextStyle;
}

const styles: Partial<BasicStyle> & AdditionalStyle = StyleSheet.create<
  Partial<BasicStyle> & AdditionalStyle
>({
  container: {
    flex: 1,
    alignContent: 'space-between',
    backgroundColor: 'white',
    flexDirection: 'row',
    paddingBottom: 8,
    paddingTop: 8,
    paddingRight: 8,
    borderBottomColor: 'lightgrey',
    borderBottomWidth: 1,
  },
  image: {
    marginRight: 16,
    marginBottom: 16,
    backgroundColor: '#e4f0f5',
  },
  imagesWrapper: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    paddingTop: 8,
  },
  tweetContainer: {
    flexShrink: 1,
    flex: 1,
  },
  text: {
    flexWrap: 'wrap',
    flexShrink: 1,
    color: '#a1a1a1',
  },
  sender: {
    color: '#4152c9',
    fontWeight: '600',
  },
  commentsWrap: {
    paddingTop: 8,
    flex: 1,
    flexDirection: 'column',
  },
  comment: {
    flex: 1,
    width: '100%',
    marginBottom: 4,
    flexDirection: 'row',
    backgroundColor: '#f7f7f7',
    padding: 2,
  },
  commentSender: {
    fontWeight: 'bold',
    color: '#000',
  },
  tweetImageContainer: {
    flex: 1,
    flexDirection: 'column',
  },
  contentWrapper: {
    flexDirection: 'column',
  },
  readMore: {
    color: 'blue',
  },
});
