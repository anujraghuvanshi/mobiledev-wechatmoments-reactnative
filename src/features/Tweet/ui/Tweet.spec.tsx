import renderer, { act } from 'react-test-renderer';
import React from 'react';

import {Tweet} from './Tweet';
import {ITweet} from '../../../types';

describe('Tweet', () => {
  const testTweet: ITweet = {
    sender: {
      nick: 'john',
      username: 'John Smith',
      avatar: 'test-image.url',
    },
    content: 'tweet content',
    images: [{url: 'image1.url'}, {url: 'image2.url'}],
    comments: [
      {
        content: 'Great tweet!',
        sender: {nick: 'user1', username: 'User One', avatar: 'avatar1.url'},
      },
      {
        content: 'I agree',
        sender: {nick: 'user2', username: 'User Two', avatar: 'avatar2.url'},
      },
    ],
  };

  it('should match component snapshots', () => {
    const component = renderer.create(<Tweet tweet={testTweet} />).toJSON();
    expect(component).toMatchSnapshot();
  });

  it('should render component', () => {
    const component = renderer.create(<Tweet tweet={testTweet} />);
    const containerElement = component.root.findByProps({
      testID: 'tweet-wrapper',
    });
    expect(containerElement).toBeTruthy();
  });

  it('should render error text on Unknown Error', () => {
    const tweet = {'unknown error': 'STARCRAFT2'};
    const component = renderer.create(<Tweet tweet={tweet} />);
    const containerElement = component.root.findByProps({
      testID: 'error-text',
    });
    expect(containerElement.props.children[1]).toBe('STARCRAFT2');
  });

  it('should render error text on normal Errors', () => {
    const tweet = {error: 'losted'};
    const component = renderer.create(<Tweet tweet={tweet} />);
    const containerElement = component.root.findByProps({
      testID: 'error-text',
    });
    expect(containerElement.props.children[1]).toEqual('losted');
  });

  it('should render the right content', () => {
    const component = renderer.create(<Tweet tweet={testTweet} />);
    const element = component.root.findByProps({
      testID: 'tweet-content',
    });
    expect(element.props.children).toBe('tweet content');
  });

  it('should verify whether read more is being called', () => {
    const component = renderer.create(<Tweet tweet={testTweet} />);
    const element = component.root.findByProps({
      testID: 'read-more-trigger',
    });
    act(() => {
      element.props.onPress();
    });
    expect(element).toHaveBeenCalled();
  });

  it("should render the sender's avatar", () => {
    const component = renderer.create(<Tweet tweet={testTweet} />);
    const avatarElement = component.root.findByProps({
      testID: 'sender-image',
    });
    expect(avatarElement.props.source.uri).toBe('test-image.url');
  });

  it("should render the sender's avatar when no image", () => {
    const tweetWithoutImages = {...testTweet, sender: {}};
    const component = renderer.create(
      <Tweet tweet={tweetWithoutImages as any} />,
    );
    const avatarElement = component.root.findByProps({
      testID: 'sender-image',
    });
    expect(avatarElement.props.source.uri).toBe('');
  });

  it('should render the images container if the tweet has images', () => {
    const component = renderer.create(<Tweet tweet={testTweet} />);
    const containerElement = component.root.findByProps({
      testID: 'tweet-images-wrapper',
    });
    expect(containerElement).toBeTruthy();
  });

  it('should render the images with blank url when tweet image is not present', () => {
    const tweetWithBlankImageObj = {...testTweet, images: [{url: null}]};
    const component = renderer.create(
      <Tweet tweet={tweetWithBlankImageObj as any} />,
    );
    const containerElement = component.root.findByProps({
      testID: 'tweet-image',
    });
    expect(containerElement.props.source.uri).toBe('');
  });

  it('should not render images container when tweet has no images', () => {
    const tweetWithoutImages = {...testTweet, images: []};
    const component = renderer.create(<Tweet tweet={tweetWithoutImages} />);
    expect(() => {
      component.root.findByProps({
        testID: 'tweet-images-wrapper',
      });
    }).toThrow();
  });

  it('should render the comments wrapper when comments exist', () => {
    const component = renderer.create(<Tweet tweet={testTweet} />);
    const commentsWrapper = component.root.findByProps({
      testID: 'tweet-comments-wrapper',
    });
    expect(commentsWrapper).toBeTruthy();
  });

  it('should not render comments wrapper when there are no comments', () => {
    const tweetWithoutComments = {...testTweet, comments: undefined};
    const component = renderer.create(<Tweet tweet={tweetWithoutComments} />);
    expect(() => {
      component.root.findByProps({
        testID: 'tweet-comments-wrapper',
      });
    }).toThrow();
  });
});
