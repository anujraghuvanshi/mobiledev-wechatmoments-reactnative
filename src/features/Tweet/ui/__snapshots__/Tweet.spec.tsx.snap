// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tweet should match component snapshots 1`] = `
<View
  style={
    Object {
      "alignContent": "space-between",
      "backgroundColor": "white",
      "borderBottomColor": "lightgrey",
      "borderBottomWidth": 1,
      "flex": 1,
      "flexDirection": "row",
      "paddingBottom": 8,
      "paddingRight": 8,
      "paddingTop": 8,
    }
  }
  testID="tweet-wrapper"
>
  <Image
    source={
      Object {
        "height": 40,
        "uri": "test-image.url",
        "width": 40,
      }
    }
    style={
      Object {
        "backgroundColor": "#e4f0f5",
        "marginBottom": 16,
        "marginRight": 16,
      }
    }
    testID="sender-image"
  />
  <View
    style={
      Object {
        "flex": 1,
        "flexShrink": 1,
      }
    }
  >
    <View>
      <Text
        style={
          Object {
            "color": "#4152c9",
            "fontWeight": "600",
          }
        }
      >
        john
      </Text>
      <View
        style={
          Object {
            "flexDirection": "column",
          }
        }
      >
        <Text
          numberOfLines={2}
          style={
            Object {
              "color": "#a1a1a1",
              "flexShrink": 1,
              "flexWrap": "wrap",
            }
          }
          testID="tweet-content"
        >
          tweet content
        </Text>
        <Text
          onPress={[Function]}
          style={
            Object {
              "color": "blue",
            }
          }
          testID="read-more-trigger"
        >
          Read more
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "column",
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "paddingTop": 8,
          }
        }
        testID="tweet-images-wrapper"
      >
        <Image
          source={
            Object {
              "height": 64,
              "uri": "image1.url",
              "width": 64,
            }
          }
          style={
            Object {
              "backgroundColor": "#e4f0f5",
              "marginBottom": 16,
              "marginRight": 16,
            }
          }
          testID="tweet-image"
        />
        <Image
          source={
            Object {
              "height": 64,
              "uri": "image2.url",
              "width": 64,
            }
          }
          style={
            Object {
              "backgroundColor": "#e4f0f5",
              "marginBottom": 16,
              "marginRight": 16,
            }
          }
          testID="tweet-image"
        />
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "column",
          "paddingTop": 8,
        }
      }
      testID="tweet-comments-wrapper"
    >
      <View
        style={
          Object {
            "backgroundColor": "#f7f7f7",
            "flex": 1,
            "flexDirection": "row",
            "marginBottom": 4,
            "padding": 2,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#000",
              "fontWeight": "bold",
            }
          }
        >
          user1
          :
           
        </Text>
        <Text>
          Great tweet!
        </Text>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#f7f7f7",
            "flex": 1,
            "flexDirection": "row",
            "marginBottom": 4,
            "padding": 2,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#000",
              "fontWeight": "bold",
            }
          }
        >
          user2
          :
           
        </Text>
        <Text>
          I agree
        </Text>
      </View>
    </View>
  </View>
</View>
`;
